src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -785,6 +785,94 @@ internal static object JoinOperator(ExecutionContext context, IScriptExtent erro
             }
         }
 
+        /// <summary>
+        /// The implementation of the PowerShell range operator.
+        /// </summary>
+        /// <param name="context">The execution context in which to evaluate the expression.</param>
+        /// <param name="errorPosition">The position to use for error reporting.</param>
+        /// <param name="lval">The object on which to start.</param>
+        /// <param name="rval">The object on which to stop.</param>
+        /// <returns>The array of objects.</returns>
+        internal static object RangeOperator(ExecutionContext context, IScriptExtent errorPosition, object lval, object rval)
+        {
+            var lbase = PSObject.Base(lval);
+            var rbase = PSObject.Base(rval);
+
+            var lType = lbase.GetType();
+            var rType = rbase.GetType();
+
+            // Return objects of [char] type for explicit [char] cast.
+            // Common case:
+            //     [char]'A'..[char]'Z'
+            // Special case:
+            //     [char]'0'..[char]'9'
+            if (lType == typeof(char) && rType == typeof(char))
```

Comment:
You can avoid the calls to GetType() and casts by using C#7 pattern matching like this:
```C#
if (lbase as char lc && rbase as char rc)
{
    return CharOps.Range(lc, rc);
}
```

The same would apply below for strings.

Reply:
Fixed.